web/blueprints/user/forms.py

```diff
@@ -42,6 +48,93 @@ def validate_unique_login(form, field):
         raise ValidationError(u"Nutzerlogin schon vergeben!")
 
 
+class UniqueName:
+    """Checks whether the name of a user in a given building is unique.
+
+    If the form misses one of the fields ``room_number``, ``level``, ``building``,
+    this validator is effectively disabled.
+
+    :param force_field: The name of the “do it anyway” checkbox field
+    """
+    def __init__(self, force_field: typing.Optional[str] = 'force'):
+        self.force_field = force_field
+        self.building_field = 'building'
+        self.level_field = 'level'
+        self.room_number_field = 'room_number'
+        self.ratio = 0.6
+
+    def force_set(self, form: Form) -> bool:
+        return self.force_field \
+               and hasattr(form, self.force_field)\
+               and getattr(form, self.force_field).data
+
+    def try_get_room(self, form: Form) -> typing.Optional[Room]:
+        try:
+            number = getattr(form, self.room_number_field).data
+            level = getattr(form, self.level_field).data
+            building = getattr(form, self.building_field).data
+        except AttributeError:
+            return
+
+        return Room.q.filter_by(number=number, level=level, building=building).one_or_none()
+
+    def similar_users(self, our_name, room: Room):
+        return [u for u in room.users
+                if SequenceMatcher(None, our_name, u.name).ratio() > self.ratio]
```

Comment:
I had a point on my todo list to improve this check. I would recommend checking for the address or the room history of the users instead of just the current room. This is probably not difficult to implement.

Reply:
I'm confused by this statement. We are talking about the uniqueness check when e.g. creating a user; before the user is created, he does not have any room history.
The only info we have is „Somebody wants to create a user with name `our_name` and move them to `room`“.

Reply:
cleared in manual conversation.

Reply:
Fixed in 71e5905eadd43ad9c503287d1decff270f1af9e8